src/gc.c

```diff
@@ -2662,7 +2662,7 @@ mark: {
             if (npointers == 0)
                 goto pop;
             uintptr_t nptr = npointers << 2 | (bits & GC_OLD);
-            assert(layout->nfields > 0 && layout->fielddesc_type != 3 && "opaque types should have been handled specially");
+            assert((layout->nfields > 0 || layout->fielddesc_type == 3) && "opaque types should have been handled specially");
```

Comment:
Just to double check my understanding of this assertion: it is triggered if `npointer > 0` yet `nfields == 0` (except for foreign types, where this is allowed). Because in this situation, the type is considered "opaque", which fits with the definition of `jl_is_layout_opaque`:
```C
static inline int jl_is_layout_opaque(const jl_datatype_layout_t *l) JL_NOTSAFEPOINT
{
    return l->nfields == 0 && l->npointers > 0;
}
```
This returns true even for foreign types -- which however makes perfect sense, as foreign types which contain npointers certainly are "opaque".


So from this perspective, this change looks correct to me.